[SNAP-3010] Cleaning block manager directories if left orphan

vatsalmevada opened the pull request:
## What changes were proposed in this pull request?
Tracking spark block manager directories for each executor and
 cleaning them in next run if left orphan.

The changes are for tracking the spark local directories (which
 are used by block manager to store shuffle data) and changes to
clean the local directories (which are left orphan due to abrupt
failure of JVM).

The changes to clean the orphan directory are also kept as part
of Spark module itself instead of cleaning it on Snappy Cluster start.
This is done because the changes to track the local directory has to
go in Spark and if the clean up is not done at the same place then
the metadata file used to track the local directories will keep 
growing while running spark cluster from snappy's spark distribution.

This handling is skipped when master is local because in local mode
driver and executors will end up writing `.tempfiles.list` file in the
same directory which may lead to concurrency issues.

## How was this patch tested?
Manual, precheckin, added a dunit test as part of https://github.com/SnappyDataInc/snappydata/pull/1377

## Other PRs
https://github.com/SnappyDataInc/snappydata/pull/1377

Commit:
Tracking spark block manager directories for each executors and
 cleaning them in next run if left orphan.

sumwale commented on core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala:
```diff
@@ -144,10 +144,12 @@ private[spark] class DiskBlockManager(conf: SparkConf, deleteFilesOnStop: Boolea
    * be deleted on JVM exit when using the external shuffle service.
    */
   private def createLocalDirs(conf: SparkConf): Array[File] = {
+    if (deleteFilesOnStop) LocalDirectoryCleanupService.clean()
```
Since this flag can change after restart, I think its better to invoke clean() without the check for deleteFilesOnStop

sumwale commented on core/src/main/scala/org/apache/spark/util/LocalDirectoryCleanupService.scala:
```diff
@@ -0,0 +1,86 @@
+   */
+  def clean(): Unit = {
+    if (Path(fileName).exists) {
+      Source.fromFile(fileName, "UTF-8").getLines().foreach(f => {
```
store Source in a val and close in a finally block

sumwale commented on core/src/main/scala/org/apache/spark/util/LocalDirectoryCleanupService.scala:
```diff
@@ -0,0 +1,86 @@
+   * @param path path to temp file/directory
+   */
+  def add(path: Path): Unit = {
+    FileUtils.writeLines(new File(fileName), "UTF-8",
```
Use File consistently instead of scala.reflect.io.Path. Alternative use java.nio.file.Path consistently that has better semantics (e.g. proper IOException for delete failure).

sumwale commented on core/src/main/scala/org/apache/spark/util/LocalDirectoryCleanupService.scala:
```diff
@@ -0,0 +1,86 @@
+   * safe manner. Also cleans the temp files list once deletion is complete.
+   */
+  def clean(): Unit = {
+    if (Path(fileName).exists) {
```
Use File.exists or java.nio.file.Path exists for consistency.

sumwale commented on core/src/main/scala/org/apache/spark/util/LocalDirectoryCleanupService.scala:
```diff
@@ -0,0 +1,86 @@
+      Source.fromFile(fileName, "UTF-8").getLines().foreach(f => {
+        try {
+          if (Path(f).exists) {
+            if (!Path(f).deleteRecursively()) {
```
Use FileUtils.deleteDirectory which will throw a better IOException for logWarning. Alternatively probably the best option is to use java.nio.file.Files.delete in conjuction with Files.walk e.g.
```scala
        Files.walk(Paths.get(localDir.getAbsolutePath))
            .sorted(Collections.reverseOrder()).forEach(new Consumer[Path] {
          override def accept(p: Path): Unit = {
            try {
              Files.delete(p)
            } catch {
              case e: Exception => // log exception
            }
          }
        })
```

The reverseOrder lists files before the containing directory to give proper deletion order. This has the advantage of logging proper exception like file in use etc which is not possible with scala/java old IO APIs or FileUtils.

sumwale commented on core/src/main/scala/org/apache/spark/util/LocalDirectoryCleanupService.scala:
```diff
@@ -0,0 +1,86 @@
+          case ex: Exception => logWarning("There was some error while deleting file: $f", ex)
+        }
+      })
+      Path(fileName).delete()
```
Like above use File.delete for consistency or Files.delete which will throw a proper IOException. If file deletion fails then it indicates some serious directory write problem in working directory so may be better to shut the JVM down with a logError followed by a System.exit like done by DiskBlockManager initialization itself in case it fails to create the directory.

sumwale commented on core/src/main/scala/org/apache/spark/util/LocalDirectoryCleanupService.scala:
```diff
@@ -0,0 +1,86 @@
+ * Contains utility methods to manage spark local directories. This service is written for handling
+ * of spark local directories left orphan due to scenario like abrupt failure of JVM.
+ */
+object LocalDirectoryCleanupService extends Logging {
```
A service in name indicates some background thread/process, so I think its better to remove Service from the name.

sumwale reviewed (commented):
See some comments

Commit:
Incorporating review comments

sumwale commented on core/src/main/scala/org/apache/spark/util/LocalDirectoryCleanupUtil.scala:
```diff
@@ -0,0 +1,103 @@
+        }
+      })
+    } else {
+      logInfo(s"File does not exists : $path")
```
s"File or directory does not exist: $path"

sumwale commented on core/src/main/scala/org/apache/spark/util/LocalDirectoryCleanupUtil.scala:
```diff
@@ -0,0 +1,103 @@
+    if (Files.exists(listFilePath)) {
+      val fileSource = Source.fromFile(listFile, "UTF-8")
+      try {
+        fileSource.getLines().map(Paths.get(_)).foreach(delete)
```
(_) is probably superfluous in Paths.get

vatsalmevada replied:
It is required as `java.nio.file.Paths#get(java.lang.String, java.lang.String...)` is having more than one arguments.

sumwale commented on core/src/main/scala/org/apache/spark/util/LocalDirectoryCleanupUtil.scala:
```diff
@@ -0,0 +1,103 @@
+          try {
+            Files.delete(p)
+          } catch {
+            case e: Exception => logError(s"Failure while deleting file: $p.", e)
```
s"Failure while deleting file or directory: $p"

sumwale commented on core/src/main/scala/org/apache/spark/util/LocalDirectoryCleanupUtil.scala:
```diff
@@ -0,0 +1,103 @@
+        Files.delete(listFilePath)
+      } catch {
+        case ex: Exception => logError(s"Failure while deleting file: $listFile.", ex)
+          System.exit(1)
```
Better to use System.exit(ExecutorExitCode.DISK_STORE_FAILED_TO_CREATE_DIR) like used by DiskBlockManager.localDirs initialization. Then the message in ExecutorExitCode.explainExitCode for this can be changed to "Failed to create or delete local ...".

vatsalmevada replied:
@sumwale At this point, we are deleting `.tempfiles.list` which is the metadata file used to track the orphan spark local dirs. Actual spark temp directories and files contained in that directory is deleted as part of  `org.apache.spark.util.LocalDirectoryCleanupUtil#delete` method. The message in `ExecutorExitCode.explainExitCode` is more specific to `spark.local.dir`. Is this suggestion actually targeted to deletiong of `spark.local.dir` and it's content?

sumwale replied:
Ok, lets keep as is for now. Perhaps a new ExecutorExitCode can be useful because driver can display the reason for exit (mesos backend handles executor exit code, and apparently we can do similar).

sumwale reviewed (approved):
minor comments

Commit:
incorporating review comments

Commit:
minor: log message correction

Commit:
Fixing failing tests-
Doing block manager directory cleanup only when Spark application is
not running in local mode. Because in local mode worker and driver
shares the JVM and end up sharing same metadata file `.tempfiles.list`

sumwale commented on core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala:
```diff
@@ -144,10 +144,15 @@ private[spark] class DiskBlockManager(conf: SparkConf, deleteFilesOnStop: Boolea
    * be deleted on JVM exit when using the external shuffle service.
    */
   private def createLocalDirs(conf: SparkConf): Array[File] = {
+    val isLocalMode = conf.contains("spark.master") && conf.get("spark.master").startsWith("local")
```
org.apache.spark.util.Utils.isLocalMaster(conf) looks standard

Commit:
incorporating review comments

Commit:
Merge branch 'snappy/branch-2.1' into SNAP-3010

Commit:
Fixing scalaStyle errors introduced by previous commits
